src/c++/perf_analyzer/test_command_line_parser.cc

```diff
@@ -0,0 +1,1040 @@
+          act = parser.Parse(argc, argv),
+          "basic_string::substr: __pos (which is 16) > this->size() (which is "
+          "15)");
+
+      optind = 1;
+    }
+
+    SUBCASE("bad shapes - a,b,c")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--shape",
+                          "input_name:a,b,c"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(parser.UsageCalled());
+      CHECK_STRING(
+          "Usage Message", parser.GetUsageMessage(),
+          "failed to parse input shape: input_name:a,b,c");
+
+      exp->input_shapes.emplace(
+          std::string("input_name"), std::vector<int64_t>{});
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("bad shapes - [1,2,3]")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--shape",
+                          "input_name:[1,2,3]"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(parser.UsageCalled());
+      CHECK_STRING(
+          "Usage Message", parser.GetUsageMessage(),
+          "failed to parse input shape: input_name:[1,2,3]");
+
+      exp->input_shapes.emplace(
+          std::string("input_name"), std::vector<int64_t>{});
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+  }
+
+  SUBCASE("Option : --measurement-interval")
+  {
+    SUBCASE("set to 500")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--measurement-interval",
+                          "500"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(!parser.UsageCalled());
+
+      exp->measurement_window_ms = 500;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("set to -200")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--measurement-interval",
+                          "-200"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(!parser.UsageCalled());
+
+      // BUG: may want to actually error out here, and not just use the unsigned
+      // conversion. This will result in unexpected behavior. The actual value
+      // becomes 18446744073709551416ULL, which is not what you would want.
+      //
+      exp->measurement_window_ms = -200;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("set to non-numeric value")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--measurement-interval",
+                          "foobar"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(parser.UsageCalled());
+      CHECK_STRING(
+          "Usage Message", parser.GetUsageMessage(),
+          "measurement window must be > 0 in msec");
+
+      exp->measurement_window_ms = 0;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+  }
+
+  // Short hand version of --measurement-interval, all test should be the same
+  //
+  SUBCASE("Option : -p")
```

Comment:
You could use `SUBCASE` like below in these situations to avoid code duplication:

https://github.com/doctest/doctest/blob/master/doc/markdown/parameterized-tests.md

```cpp
SUBCASE("Measurement interval set to 500") {
  int argc = 5;
  char *argv[argc] = {app_name, "-m", model_name, "", "500"};

  SUBCASE("Option : --measurement-interval") {
    argv[3] = "--measurement-interval";
  }
  SUBCASE("Option : -p") { argv[3] = "-p"; }

  CAPTURE(argv[3]);

  // testing logic for '500' case
}

SUBCASE("Measurement interval set to -200") {
  int argc = 5;
  char *argv[argc] = {app_name, "-m", model_name, "", "-200"};

  SUBCASE("Option : --measurement-interval") {
    argv[3] = "--measurement-interval";
  }
  SUBCASE("Option : -p") { argv[3] = "-p"; }

  CAPTURE(argv[3]);

  // testing logic for '-200' case
}

SUBCASE("Measurement interval set to non-numeric value") {
  int argc = 5;
  char *argv[argc] = {app_name, "-m", model_name, "", "foobar"};

  SUBCASE("Option : --measurement-interval") {
    argv[3] = "--measurement-interval";
  }
  SUBCASE("Option : -p") { argv[3] = "-p"; }

  CAPTURE(argv[3]);

  // testing logic for non-numeric value case
}
```

Reply:
I will look into that.

Reply:
Are you intending to remove the mentioned unit testing code duplication in this PR?

Reply:
This was a good suggestion, I hadn't realized doctest allows you to nest unit tests in this manner. I will use this moving forward. New code removes the duplication.